designer/server/src/main/scala/pl/touk/nussknacker/ui/process/ProcessStateDefinitionService.scala

```diff
@@ -0,0 +1,79 @@
+package pl.touk.nussknacker.ui.process
+
+import io.circe.generic.JsonCodec
+import pl.touk.nussknacker.engine.ProcessingTypeData
+import pl.touk.nussknacker.engine.api.deployment.StateDefinitionDetails
+import pl.touk.nussknacker.engine.api.deployment.StateStatus.StatusName
+import pl.touk.nussknacker.restmodel.process.ProcessingType
+import pl.touk.nussknacker.ui.process.ProcessStateDefinitionService.processingTypeStateDefinitions
+import pl.touk.nussknacker.ui.process.processingtypedata.ProcessingTypeDataProvider
+
+import java.net.URI
+
+class ProcessStateDefinitionService(typeToConfig: ProcessingTypeDataProvider[ProcessingTypeData],
+                                    categoryService: ProcessCategoryService) {
+
+  def fetchStateDefinitions(): List[UIStateDefinition] = {
+    processingTypeStateDefinitions(typeToConfig.all)
+      .groupBy { case (_, statusName, _) => statusName }
+      .map { case (statusName, stateDefinitionsByName) =>
+        // here we assume it is asserted that all definitions are unique
+        val stateDefinition = stateDefinitionsByName.map { case (_, _, sd) => sd }.head
+        val categoriesWhereStateAppears = stateDefinitionsByName
+          .flatMap { case (processingType, _, _) => categoryService.getProcessingTypeCategories(processingType) }
+        UIStateDefinition(
+          statusName,
+          stateDefinition.displayableName,
+          stateDefinition.icon,
+          stateDefinition.tooltip,
+          categoriesWhereStateAppears
+        )
+      }
+      .toList
+  }
+}
+
+// TODO: Move validation and stateDefinitions aggregation to ProcessingTypeDataReader (where processing types are reloaded).
+// Now we have only late validation when UI asks for state definitions, and state aggregation is performed here
+// every time validation is executed, but should happen only once per configuration reload.
+// That requires ProcessingTypeDataProvider.all to be an entity that combines all processing types AND
+// has all-processing-type aggregates, such as aggregated stateDefinitions.
+object ProcessStateDefinitionService {
+
+  /**
+    * Each processing type define its own state definitions. Technically it is possible that two processing types provide
+    * states with the same StatusName and different UI configurations (displayable name and icon). Here is an assertion
+    * that this does not happen and each state has the same definition across all processingTypes.
+    */
+  def validate(processingTypes: Map[ProcessingType, ProcessingTypeData]): Unit = {
```

Comment:
Hmm I think it should return Validated[Ex, Unit]?

Reply:
Or rather change the name to `checkUnsafe` It meant to fail with an exception. It is rather a constraints violation check, not business parameters validation.

Reply:
I've changed the name to `checkUnsafe`.